generative/maisi/scripts/utils.py

```diff
@@ -0,0 +1,394 @@
+        ),
+        0,
+        1,
+    )
+
+
+def supress_non_largest_components(img, target_label, default_val=0):
+    """As a last step, supress all non largest components"""
+    index_arr = get_index_arr(img)
+    img_mod = copy.deepcopy(img)
+    new_background = np.zeros(img.shape, dtype=np.bool_)
+    for label in target_label:
+        label_cc = skimage.measure.label(img == label, connectivity=3)
+        uv, uc = np.unique(label_cc, return_counts=True)
+        dominant_vals = uv[np.argsort(uc)[::-1][:2]]
+        if len(dominant_vals) >= 2:  # Case: no predictions
+            new_background = np.logical_or(
+                new_background,
+                np.logical_not(np.logical_or(label_cc == dominant_vals[0], label_cc == dominant_vals[1])),
+            )
+
+    for voxel in index_arr[new_background]:
+        img_mod[tuple(voxel)] = default_val
+    diff = np.sum((img - img_mod) > 0)
+
+    return img_mod, diff
+
+
+def erode3d(input_tensor, erosion=3, value=0.0):
+    # Define the structuring element
+    erosion = ensure_tuple_rep(erosion, 3)
+    structuring_element = torch.ones(1, 1, erosion[0], erosion[1], erosion[2]).to(input_tensor.device)
+
+    # Pad the input tensor to handle border pixels
+    input_padded = F.pad(
+        input_tensor.float().unsqueeze(0).unsqueeze(0),
+        (erosion[0] // 2, erosion[0] // 2, erosion[1] // 2, erosion[1] // 2, erosion[2] // 2, erosion[2] // 2),
+        mode="constant",
+        value=value,
+    )
+
+    # Apply erosion operation
+    output = F.conv3d(input_padded, structuring_element, padding=0)
+
+    # Set output values based on the minimum value within the structuring element
+    output = torch.where(output == torch.sum(structuring_element), 1.0, 0.0)
+
+    return output.squeeze(0).squeeze(0)
+
+
+def dilate3d(input_tensor, erosion=3, value=0.0):
+    # Define the structuring element
+    erosion = ensure_tuple_rep(erosion, 3)
+    structuring_element = torch.ones(1, 1, erosion[0], erosion[1], erosion[2]).to(input_tensor.device)
+
+    # Pad the input tensor to handle border pixels
+    input_padded = F.pad(
+        input_tensor.float().unsqueeze(0).unsqueeze(0),
+        (erosion[0] // 2, erosion[0] // 2, erosion[1] // 2, erosion[1] // 2, erosion[2] // 2, erosion[2] // 2),
+        mode="constant",
+        value=value,
+    )
+
+    # Apply erosion operation
+    output = F.conv3d(input_padded, structuring_element, padding=0)
+
+    # Set output values based on the minimum value within the structuring element
+    output = torch.where(output > 0, 1.0, 0.0)
+
+    return output.squeeze(0).squeeze(0)
+
+
+def organ_fill_by_closing(data, target_label, device):
+    mask = (data == target_label).astype(np.uint8)
+    mask = dilate3d(torch.from_numpy(mask).to(device), erosion=3, value=0.0)
+    mask = erode3d(mask, erosion=3, value=0.0)
+    mask = dilate3d(mask, erosion=3, value=0.0)
+    mask = erode3d(mask, erosion=3, value=0.0).cpu().numpy()
```

Comment:
Perhaps we can refactor it into a closing function which can introduce an iteration number and avoid hard-coding erosion here. What do you think?

Reply:
Done, thanks